### PR TITLE
Add sidekiq gem

### PR DESCRIPTION
**Pull request by ViditChitkara**
part of #2104 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [ ] code is in uniquely-named feature branch and has no merge conflicts
* [ ] PR is descriptively titled
* [ ] PR body includes `fixes #0000`-style reference to original issue #
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. 

Thanks!

### CODE REVIEW

**Comment by ViditChitkara** [comment]
@icarito @jywarren if we could start with this?

**Comment by jywarren** [comment]
@icarito does this look good? Thanks Vidit!

**Comment by icarito** [comment]
From https://stackoverflow.com/questions/285717/a-cron-job-for-rails-best-practices : 

_"(...) Keep in mind that Resque is a dead project (not actively maintained), so Sidekiq is a way better alternative. It also is more performant: Sidekiq runs several workers on a single, multithread process while Resque runs each worker in a separate process._

edited Sep 16 '15 at 18:23
answered Oct 24 '14 at 20:20 "

@jywarren you have a much deeper understanding of Gem ecosystem and plots2 in particular so those are just my comments as I try to keep us running only on actively maintained dependencies. Feel free to merge if you think it's fine!

**Comment by jywarren** [comment]
Vidit, can you add some extra info to help guide the next step? Thanks!

**Comment by ViditChitkara** [comment]
Hi @icarito , Yes we need redis service too! Redis service is mandatory for resque. I think we need to manually install redis on our production servers. I also did manual installation of redis (it is a key-value based database) on my local environment for testing out resque. So, maybe some similar method must be there for redis installation on the server.
`Whenever` gem just executes the file or command specified in the scheduler.
So these are two things through which we could accomplish recurring mails:-
1) Whenever gem - specifies just time and date on which the specified file runs
2) Resque+redis - The file which runs are en-queued here and executed asynchronously.

**Comment by ViditChitkara** [comment]
> Vidit, can you add some extra info to help guide the next step? Thanks!
Trying to dig out some resources. Will update here soon.

**Comment by ViditChitkara** [comment]
https://stackoverflow.com/questions/23265242/redis-in-ubuntu
https://redis.io/download
https://www.digitalocean.com/community/tutorials/how-to-install-and-configure-redis-on-ubuntu-16-04
@icarito we can get help from these I guess.

**Comment by jywarren** [comment]
@ViditChitkara - I think @icarito is saying that resque is no longer actively supported and suggesting sidekiq as a possible alternative. 

I'm seeing last update 2 days ago for https://github.com/mperham/sidekiq/pulse/monthly

17 days ago for resque: https://github.com/resque/resque/pulse/monthly

Since Redis says it's compatible with both, what do you think?

**Comment by jywarren** [comment]
@icarito how shall we proceed on getting redis installed -- i believe you'd mentioned a container somewhere -- can we link that thread in here? 

Let's really try using the `unstable` branch to try a lot of this before merging to master, ok?

**Comment by jywarren** [comment]
@ViditChitkara if you can link us to the past conversation (try searching for it?) about containers and redis -- that'll help us plot out the next steps. Thanks!

**Comment by icarito** [comment]
The general idea is that we'll need to add Redis to `docker-compose.yml` file. We should get https://github.com/publiclab/plots2/pull/2593 merged and then we can get working on adding Redis service to the appropriate docker-compose file on each instance.

**Comment by icarito** [comment]
If we can merge the Gem and then cofigure the link to a Redis service that'd be fine from my side. That way we can get this merged and make another PR with the required changes to the deployment process and container definitions.

**Comment by ViditChitkara** [comment]
> If we can merge the Gem and then cofigure the link to a Redis service that'd be fine from my side. That way we can get this merged and make another PR with the required changes to the deployment process and container definitions.

@icarito so should be currently stick to resque or should we use sidekiq??

**Comment by ViditChitkara** [comment]
@icarito @jywarren I have added sidekiq gem. Please have a look!

**Comment by namangupta01** [comment]
@ViditChitkara There is code climate error, please fix it. Then it will be good to go.

**Comment by icarito** [comment]
The redis container image to be integrated is https://hub.docker.com/_/redis/

**Comment by ViditChitkara** [comment]
Thanks for your guidance @icarito . I did some research regarding docker and docker-compose and made some changes. However, travis tests are failing, could you please have a look, where it went wrong?
Thanks!!

**Comment by namangupta01** [comment]
@ViditChitkara Same error is happening in #2669. I think there is some problem in the master itself.

**Comment by icarito** [comment]
@namangupta01 reported the same issue last night I think it has something to do with octokit Gem. https://github.com/octokit/octokit.rb

**Comment by icarito** [comment]
It is failing on a Github API endpoint. Look at this pull request https://github.com/octokit/octokit.rb/pull/1016/commits/491476a97e52ea77babab40303a5f01f0a94cf27 possibly they have changed their API or something. But I wonder why our build process depends on the Github API ?

**Comment by Gauravano** [comment]
Facing the same issue at #2747

**Comment by jywarren** [comment]
This looks to be an issue with the Danger bot parsing of errors and reporting back the errors in the PR -- which needs GitHub access:

```
/home/travis/.rvm/gems/ruby-2.4.1/gems/octokit-4.9.0/lib/octokit/response/raise_error.rb:16:in `on_complete': GET https://api.github.com/repos/publiclab/plots2/pulls/2738: 401 - Bad credentials // See: https://developer.github.com/v3 (Octokit::Unauthorized)
```

Can someone look at Danger's Ruby issue tracker to see if we need to do something to get Danger running again?

**Comment by jywarren** [comment]
http://danger.systems

**Comment by jywarren** [comment]
@namangupta01 is looking at this too -- and i just looked around a bit --

Could it have to do with https://github.com/danger/danger/issues/984 and 
https://developer.github.com/changes/2018-05-07-new-checks-api-public-beta/ ??

**Comment by jywarren** [comment]
I think this is the Octokit gem we're using -- i don't see an obvious sign that something with the credentials has changed... https://github.com/octokit/octokit.rb#making-requests

**Comment by jywarren** [comment]
We should check that our token works with https://github.com/danger/danger/issues/675#issuecomment-265448151

**Comment by jywarren** [comment]
OK, i tried also revoking and re-accepting TravisCI's access to the repository in GitHub.

**Comment by jywarren** [comment]
Retriggered it. Let's see -- if this doesn't work, any help appreciated in figuring out what's going wrong with the `octokit` gem credentials... we may need to open an issue on the Danger tracker...

**Comment by ViditChitkara** [comment]
Hi all,
@jywarren , @icarito Regarding the octokit error, someone else has faced "unauthorised" issue too! Have a look at this [pr](https://github.com/travis-ci/travis-ci/issues/9610).
Digging out some other resources to fix this.

**Comment by jywarren** [comment]
oh great find, Vidit! We're all quite stuck here :-P

**Comment by ViditChitkara** [comment]
Well, working to find out it's fix!! Hope we get one!

**Comment by jywarren** [comment]
Tracking the octokit/travis bug now in this issue: https://github.com/publiclab/plots2/issues/2758 Please pitch in!

**Comment by jywarren** [comment]
🎉

**Comment by jywarren** [comment]
Hi, Vidit - can you update us on the status on this PR now? Thanks!

**Comment by ViditChitkara** [comment]
Resbased, the tests are running!! Let's see!!

**Comment by jywarren** [comment]
Awesome. Is this ready to be merged and pushed to production? Or do we need something else @icarito @viditchitkara?

**Comment by jywarren** [comment]
And perhaps any changes due to codeclimate recommendation?

**Comment by ViditChitkara** [comment]
Ready according to me!! Let's wait for @icarito 's call!!

**Comment by icarito** [comment]
Hi @ViditChitkara ! Great to see this moving forward.
We still need to add a working Redis setup to `docker-compose.yml`. I left review notes. Only a single container `redis` is needed. Do you have Docker available in your development environment to be able to test it?

If not, to keep this in small PR, we can merge if we make sure it doesn't break things when no Redis service is available, and work on adding the Redis service in a different PR.

File: docker-compose.yml [hunk 1]
```diff
@@ -16,6 +16,8 @@ services:
       - ../dump:/docker-entrypoint-initdb.d
       # for importing a dump
       # if you want to re-import simply put away ../mysql/*
+  redis:
+    image: redis:latest
```
Comment (ViditChitkara) [inline docker-compose.yml]: hi @icarito , I have added redis image here. Does it require anything else?
I have tested it locally using `docker-compose up` and it went well I guess. Please have a look!!
Thanks!!

**Comment by jywarren** [comment]
The CodeClimate issues are looking for this kind of change to redirects: `redirect_to URI.parse("/string-url-here").path` -- would you mind fixing a few of those? They may not be related but we're gradually fixing these across the site. Thanks!

**Comment by ViditChitkara** [comment]
Yeah I'll do!!

**Comment by ViditChitkara** [comment]
The conflicts are resolved!! Thanks for the help @jywarren .

**Comment by ViditChitkara** [comment]
Please have a look at the redis container code in docker-compose file. If that's good to go, we could merge this.

**Comment by jywarren** [comment]
Great -- if @icarito gives us a green light on the `redis` setup, we can merge! @icarito go ahead if this looks ready!

**Comment by icarito** [comment]
Hi @ViditChitkara and @jywarren - The `docker-compose.yml` file is wrong because only *one* new container is needed (`redis`) - Vidit added another container called `sidekiq` which is not needed at all. Please remove this extra container.

**Comment by ViditChitkara** [comment]
Hi @icarito, redis alone won't help, we need to start sidekiq (bundle exec sidekiq) and maybe that's why we need to create an extra container for this. Saw various examples, have a look at:-
https://stackoverflow.com/questions/33563161/how-to-put-sidekiq-into-docker-in-a-rails-application

**Comment by ViditChitkara** [comment]
Also, do I need to shift the env variable to web container, I mean append to that environment list??

**Comment by icarito** [comment]
Oh sorry then @ViditChitkara I misunderstood. I thought we needed one schedule but it seems we need an extra process to schedule the scheduler? @jywarren this feels like an overly complex setup for scheduling but I guess it's common practice - I'm fine with merging in that case.

**Comment by ViditChitkara** [comment]
Cool, let's see how this works after merging, I guess it's good to go then!! @icarito, @jywarren.

**Comment by jywarren** [comment]
Awesome, merged! How will we confirm it works? 

Thanks all!

**Comment by ViditChitkara** [comment]
Don't really know how to check if it is running on production. However I will create a route to mount sidekiq server where we could view all the queues and workers on a separate portal.

**Comment by icarito** [comment]
Vidit, congratulations on a milestone that we do need ! :-)  :1st_place_medal: 
I was just now reading up on GitLab and found [a note](https://docs.gitlab.com/ee/administration/operations/sidekiq_memory_killer.html) that there is a possibility to leak memory with Sidekiq but according to its developers [it depends a lot on how we use it](https://github.com/mperham/sidekiq/wiki/Problems-and-Troubleshooting).

@jywarren do please review that last link I just shared because it lists some Gems that we do use as potentially problematic: RMagick, libxml-ruby and therubyracer sound familiar from our Gemfile.

If memory consumption grows, stability and performance can suffer so let's keep this in mind please! Thanks!

**Comment by jywarren** [comment]
We definitely use `therubyracer`. I think this probably merits its own issue, esp since this is closed now. @ViditChitkara can you open one? 

Is this an absolute requirement not to use `therubyracer` and `rmagick`? Or just not in conjunction with `sidekiq`? Let's make a checklist of items we need to replace. 

What will happen if we are running production code without these gems removed? We'll see instability once memory runs out?

**Comment by jywarren** [comment]
If you can copy in this info:

https://github.com/sstephenson/execjs/issues/205#issuecomment-371472473 says `therubyracer` can be a problem with `sidekiq` -- but `execjs` can be run with `nodejs` instead. That seems like a way forward. I'm not sure what's relying on `therubyracer` here, actually - it's been in the repo a long time!

**Comment by jywarren** [comment]
It does look like `execjs` and `therubyracer` are only used to compile assets? https://github.com/publiclab/plots2/blob/master/Gemfile#L44

**Comment by jywarren** [comment]
`RMagick` isn't installed -- `paperclip` just uses `imageMagick` directly outside Ruby.

**Comment by jywarren** [comment]
and `libxml-ruby` also not in Gemfile. Maybe we're ok?

**Pull request merged**